In [ ]:
import numpy as np
import pandas as pd
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression, Lasso
from sklearn.model_selection import KFold
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.impute import SimpleImputer
import matplotlib.pyplot as plt


In [ ]:
df1=pd.read_csv('train.csv')
df2=pd.read_csv('test.csv')
X_train=df1[['

In [ ]:
df1.head()

In [ ]:
df1.info()

In [ ]:
df2.info()

In [ ]:
df1.describe()
df1.isna().sum().sort_values(ascending=False)

In [ ]:
df1.isna().sum().sort_values(ascending=False)

In [ ]:
Xtrain=df1[['Name','Pclass','Age','Sex','Parch','SibSp','Fare','Survived']]
Xtrain.info()

In [ ]:
Xtrain.isna().sum().sort_values(ascending=False)

the missing data is around 20% of total data, so cant just delete(remove) the missing values. My approch will be to impute data and add mean of the ages to the missing places

In [ ]:
imp_age_train=SimpleImputer(strategy='mean')
Xtrain['Age']=imp_age_train.fit_transform(Xtrain[['Age']])

In [ ]:
Xtrain.isna().sum().sort_values(ascending=False)

In [ ]:
Xtest=df2[['Name','Pclass','Age','Sex','Parch','SibSp','Fare']]
Xtest.info()
Xtest.isna().sum().sort_values(ascending=False)

here we will impute the age and remove 1 fare record

In [ ]:
imp_age_test=SimpleImputer(strategy='mean')
Xtest=Xtest.dropna(subset=['Fare'])
Xtest['Age']=imp_age_test.fit_transform(Xtest[['Age']])


In [ ]:
Xtest.isna().sum().sort_values(ascending=False)

In [ ]:
Xtrain=Xtrain.drop('Sex', axis=1).join(pd.get_dummies(Xtrain['Sex'], drop_first=True).astype(int))
Xtest=Xtest.drop('Sex', axis=1).join(pd.get_dummies(Xtest['Sex'], drop_first=True).astype(int))


In [ ]:
Xtrain.head()

KNeighborsClassifier

In [158]:
for k in range(1, 11):
     knn=KNeighborsClassifier(n_neighbors=k)
     y=Xtrain['Survived']
     X=Xtrain.drop(['Survived','Name'], axis=1)
     knn.fit(X,y)
     y_pred=knn.predict(X)
     print('Training Accuracy for k=',k,':', knn.score(X,y))
print(Xtrain['Name'])

Training Accuracy for k= 1 : 0.9775533108866442
Training Accuracy for k= 2 : 0.8462401795735129
Training Accuracy for k= 3 : 0.8395061728395061
Training Accuracy for k= 4 : 0.8058361391694725
Training Accuracy for k= 5 : 0.8114478114478114
Training Accuracy for k= 6 : 0.7957351290684624
Training Accuracy for k= 7 : 0.7822671156004489
Training Accuracy for k= 8 : 0.7732884399551067
Training Accuracy for k= 9 : 0.7744107744107744
Training Accuracy for k= 10 : 0.7631874298540965
0                                Braund, Mr. Owen Harris
1      Cumings, Mrs. John Bradley (Florence Briggs Th...
2                                 Heikkinen, Miss. Laina
3           Futrelle, Mrs. Jacques Heath (Lily May Peel)
4                               Allen, Mr. William Henry
                             ...                        
886                                Montvila, Rev. Juozas
887                         Graham, Miss. Margaret Edith
888             Johnston, Miss. Catherine Helen "Carrie"
889   

k=1: 97.8%  ← Overfitting (memorizes training data)
k=2-5: 81-84% ← Good balance  
k=6+: 76-78%  ← Underfitting (too generalized)